langchain/retrievers/self_query/pinecone.py

```diff
@@ -0,0 +1,56 @@
+""""""
+from typing import Dict, List, Tuple, Union
+
+from langchain.chains.query_constructor.query_language import (
+    Comparator,
+    Comparison,
+    Operation,
+    Operator,
+    StructuredQuery,
+    Visitor,
+)
+
+
+class PineconeTranslator(Visitor):
+    allowed_operators = [Operator.AND, Operator.OR]
+
+    def _format_func(self, func: Union[Operator, Comparator]) -> str:
+        """"""
+        if isinstance(func, Operator) and self.allowed_operators is not None:
+            if func not in self.allowed_operators:
+                raise ValueError
+        if isinstance(func, Comparator) and self.allowed_comparators is not None:
+            if func not in self.allowed_comparators:
+                raise ValueError
+        return f"${func}"
+
+    def visit_operation(self, operation: Operation) -> Dict:
```

Comment:
A recursive solution is more natural here -- it incurs some additional overhead but unlikely to be a bottleneck for us given the context (llm invocations consume significantly more resources)

With a recursive variant I think this code becomes a single line:

```python
def visit_operation(operation):
   return {
       self.format_func(operation.operator): [ arg.accept(self) for arg in operation.arguments ]
    }
```